tests/functional_tests/ffront_tests/test_program.py

```diff
@@ -0,0 +1,291 @@
+        match=(r"Invalid Program Syntax: Missing symbol definitions: {'out_field'}"),
+    ):
+        ProgramParser.apply_to_function(undefined_field_program)
+
+
+@pytest.mark.xfail
+def test_inout_prohibited():
+    def inout_field_program(inout_field: Field[[IDim], "float64"]):
+        identity(inout_field, out=inout_field)
+
+    with pytest.raises(
+        GTTypeError,
+        match=(r"Call to function with field as input and output not allowed."),
+    ):
+        ProgramLowering.apply(ProgramParser.apply_to_function(inout_field_program))
+
+
+def invalid_call_sig_program():
+    with pytest.raises(
+        GTTypeError,
+    ) as exc_info:
+        ProgramLowering.apply(ProgramParser.apply_to_function(invalid_call_sig_program))
+
+    assert exc_info.match("Invalid call to `identity`")
+    assert exc_info.match("Function takes 1 arguments, but 2 were given.")
+    assert exc_info.match("Missing required keyword argument\(s\) `out`")
+
+
+def test_copy_restrict_parsing():
+    past_node = ProgramParser.apply_to_function(copy_restrict_program.definition)
+
+    field_type = common_types.FieldType(
+        dims=[IDim],
+        dtype=common_types.ScalarType(kind=common_types.ScalarKind.FLOAT64, shape=None),
+    )
+    slice_pattern_node = past_.Slice(lower=past_.Constant(value=1), upper=past_.Constant(value=2))
+    pattern_node = past_.Program(
+        id="copy_restrict_program",
+        params=[
+            past_.Symbol(id="in_field", type=field_type),
+            past_.Symbol(id="out_field", type=field_type),
+        ],
+        body=[
+            past_.Call(
+                func=past_.Name(id="identity"),
+                args=[past_.Name(id="in_field")],
+                kwargs={
+                    "out": past_.Subscript(
+                        value=past_.Name(id="out_field"), slice_=slice_pattern_node
+                    )
+                },
+            )
+        ],
+    )
+
+    pattern_node.matches(past_node, raise_=True)
+
+
+def test_copy_lowering():
+    past_node = ProgramParser.apply_to_function(copy_restrict_program.definition)
+    itir_node = ProgramLowering.apply(past_node)
+    closure_pattern = itir_.StencilClosure(
+        domain=itir_.FunCall(
+            fun=itir_.SymRef(id="domain"),
+            args=[
+                itir_.FunCall(
+                    fun=itir_.SymRef(id="named_range"),
+                    args=[
+                        itir_.AxisLiteral(value="IDim"),
+                        itir_.IntLiteral(value=1),
+                        itir_.IntLiteral(value=2),
+                    ],
+                )
+            ],
+        )
+    )
+    fencil_pattern = itir_.FencilDefinition(
+        id="copy_restrict_program",
+        params=[
+            itir_.Sym(id="in_field"),
+            itir_.Sym(id="out_field"),
+            itir_.Sym(id="__in_field_size_0"),
+            itir_.Sym(id="__out_field_size_0"),
+        ],
+        closures=[closure_pattern],
+    )
```

Comment:
This goes for all Iterator IR patterns: They all seem complete, as in every attribute is given (ITIR being very minimal this is not surprising). Is there any reason for using the pattern matching here, except to avoid false negatives in the future if things like source locations should be added to ITIR?

Reply:
False negatives in the future were one reason, but the main reason is that the `StencilClosure` is not complete here. There is now an additional test, that is actually testing each and every attribute, but I still use the pattern matching to avoid having a different way of comparing there (the pattering matching could be improved in that respect).